python/mxnet/gluon/block.py

```diff
@@ -1054,34 +1098,16 @@ def register_op_hook(self, callback, monitor_all=False):
     def forward(self, x, *args):
         """Defines the forward computation. Arguments can be either
         :py:class:`NDArray` or :py:class:`Symbol`."""
-        flatten_args = _flatten([x] + list(args), 'inputs')[0]
-        is_ndarray = None
-        ctx = None
-        exist_sym_nd = False
-        for ele in flatten_args:
-            if isinstance(ele, NDArray):
-                if is_ndarray is False:
-                    raise ValueError('In HybridBlock, we do not support mixed NDArrays and Symbols'
-                                     ' types for the input.\n'
-                                     'Received types are: {}.'
-                                     .format([type(ele) for ele in flatten_args]))
-                is_ndarray = True
-                exist_sym_nd = True
-                ctx = ele.context
```

Comment:
This line introduced in #16280 is not compatible with the previous context handling. Previously, always `x.context` is used as default context. https://github.com/apache/incubator-mxnet/pull/16280/files#diff-29da832c2145752f3906a2f71c7b63baL982
Does it matter?

Reply:
It is chosen like this because x can be None now.

Reply:
Then it should be set to the first non-`None` argument, not the last?

Reply:
Actually, all ctxs are supposed to be the same. For example, we should not allow the mixing of cpu and gpu contexts. However, we currently allow to do so because we will need to mix  `cpu`, `cpu_pinned`, and `cpu_shared`.

Reply:
Thus we should use the first non-`None` argument not to break backwards compatibility? ` cpu, cpu_pinned, cpu_shared` are different contexts after all

Reply:
@leezu I think the difference between using the first or last non-None argument is not very important here. We need to just make sure that we will finally pick a meaningful one. In fact, the previous implementation has not checked the ctx of all the args.

Reply:
As the previous implementation hasn't enforced all contexts being equal, we shouldn't start picking a different array to determine the context. As you stated above, it's valid to use a mix of  `cpu, cpu_pinned, cpu_shared` contexts.
For example, after your change, `cpu_pinned` or `cpu_shared` may be picked as default context instead of `cpu` if the user passed a  `cpu_pinned` or `cpu_shared`  as last argument. The extra overhead could cause a performance regression (all parameters will be made available under default context).
No need to risk this given there is no advantage?

Reply:
@leezu It's also possible that, previously cpu_pinned is picked as the default argument and after the change, the correct cpu context is picked as the default. My point is we need to probably give special treatment of the `cpu, cpu_pinned, cpu_shared`. What's your opinion?

Reply:
@leezu I agree that the backward-compatible issue is valid. Let me first make it to be backward-compatible. However, this does not fix the issue of the cpu, cpu_pinned, cpu_shared combination.

Reply:
We should get rid of choosing one array and using it's context as default context. For parameters, users should get the array via `self.weight.data(ctx)`. For the time being I suggest not to break the behaviour, to avoid unintended consequences